data/decode.go

```diff
@@ -0,0 +1,554 @@
+		if prefix == "" {
+			// time.Time or Timestamp is passed directly to Decode. They have
+			// to be a field of a struct.
+			return errors.New("timestamp cannot be decoded directly")
+		}
+		return d.decodeTimestamp(prefix, src, dst)
+	}
+
+	m, err := AsMap(src)
+	if err != nil {
+		return fmt.Errorf("%v: struct can only be decoded from a map", prefix)
+	}
+
+	errPrefix := prefix + "."
+	if prefix == "" {
+		errPrefix = ""
+	}
+
+	var unused map[string]struct{}
+	if d.config.ErrorUnused {
+		unused = make(map[string]struct{}, len(m))
+		for k := range m {
+			unused[k] = struct{}{}
+		}
+	}
+
+	// Accumulate all error informations to help users debug BQL.
+	errs := d.iterateField(errPrefix, m, unused, dst)
+	if d.config.ErrorUnused && len(unused) > 0 {
+		keys := make([]string, len(unused))
+		i := 0
+		for k := range unused {
+			keys[i] = errPrefix + k
+			i++
+		}
+		if d.config.Metadata != nil {
+			d.config.Metadata.Unused = append(d.config.Metadata.Unused, keys...)
+		}
+
+		errs = multierror.Append(errs, fmt.Errorf("%v: unused keys: %v", prefix, strings.Join(keys, ", ")))
+	}
+	if errs == nil {
+		// To avoid nil != nil problem due to type mismatch, this function has
+		// to return nil explicitly. Don't do "return errs" even if errs == nil.
+		return nil
+	}
+
+	// TODO: set formatter
+	return errs
+}
+
+func (d *Decoder) iterateField(prefix string, m Map, unused map[string]struct{}, dst reflect.Value) *multierror.Error {
+	// FIXME: iterateField decodes the same value multiple times if a struct and
+	// its embedded struct have the same field names.
+
+	// Accumulate all error informations to help users debug BQL.
+	var errs *multierror.Error
+
+	if dst.Type().ConvertibleTo(reflect.TypeOf(time.Time{})) {
+		errs = multierror.Append(errs, fmt.Errorf("%v: time.Time and data.Timestamp cannot be embedded", prefix))
+		return errs
+	}
+
+	t := dst.Type()
+	for i, n := 0, t.NumField(); i < n; i++ {
+		f := t.Field(i)
+		if f.Anonymous { // process embedded field
+			if f.Type.Kind() == reflect.Struct {
+				if err := d.iterateField(prefix, m, unused, dst.Field(i)); err != nil {
+					errs = multierror.Append(errs, err)
+				}
+				continue
+
+			} else if f.Type.Kind() == reflect.Ptr && f.Type.Elem().Kind() == reflect.Struct {
+				v := reflect.New(f.Type.Elem())
+				if err := d.iterateField(prefix, m, unused, reflect.Indirect(v)); err != nil {
+					errs = multierror.Append(errs, err)
+					continue
+				}
+				dst.Field(i).Set(v)
+				continue
+			}
+			errs = multierror.Append(errs, fmt.Errorf("%v: unsupported embedded field: %v", prefix, f.Name))
+		}
+
+		opts := strings.Split(f.Tag.Get(d.config.TagName), ",")
+
+		// parse options
+		var (
+			required    bool
+			weaklyTyped bool
+		)
+		for ti, opt := range opts {
+			if ti == 0 { // skip name
+				continue
+			}
+			switch opt {
+			case "required":
+				required = true
+			case "weaklytyped":
+				weaklyTyped = true
+			default:
```

Comment:
I fixed it to return a special error message when the option is an empty string.